lib/fencing/st_client.c

```diff
@@ -749,19 +749,41 @@ update_remaining_timeout(stonith_action_t * action)
 int
 stonith__result2rc(const pcmk__action_result_t *result)
 {
+    if (pcmk__result_ok(result)) {
+        return pcmk_rc_ok;
+    }
+
     switch (result->execution_status) {
+        case PCMK_EXEC_PENDING:         return EINPROGRESS;
         case PCMK_EXEC_CANCELLED:       return ECANCELED;
         case PCMK_EXEC_TIMEOUT:         return ETIME;
         case PCMK_EXEC_NOT_INSTALLED:   return ENOENT;
         case PCMK_EXEC_NOT_SUPPORTED:   return EOPNOTSUPP;
         case PCMK_EXEC_NOT_CONNECTED:   return ENOTCONN;
         case PCMK_EXEC_NO_FENCE_DEVICE: return ENODEV;
         case PCMK_EXEC_NO_SECRETS:      return EACCES;
-        default:                        break;
-    }
 
-    if (pcmk__result_ok(result)) {
-        return pcmk_rc_ok;
+        /* For the fencing API, PCMK_EXEC_INVALID is used with fencer API
+         * operations that don't involve executing an agent (for example,
+         * registering devices). This allows us to use the CRM_EX_* codes in the
+         * exit status for finer-grained responses.
+         */
+        case PCMK_EXEC_INVALID:
+            switch (result->exit_status) {
+                case CRM_EX_INSUFFICIENT_PRIV:  return EACCES;
+                case CRM_EX_PROTOCOL:           return EPROTO;
+
+               /* CRM_EX_EXPIRED is used for orphaned fencing operations left
+                * over from a previous instance of the fencer. For API backward
+                * compatibility, this is mapped to the previously used code for
+                * this case, EHOSTUNREACH.
+                */
+                case CRM_EX_EXPIRED:            return EHOSTUNREACH;
+                default:                        break;
+            }
+
```

Comment:
maybe a break; or fallthrough remark here ...

Reply:
return/break is to the right (not our usual style but I think it's clearer for these mapping switches). If it's confusing though I can change it to the usual style

Reply:
was referring to the outer switch-case

Reply:
was referring to the outer switch-case

Reply:
whoops, i just missed that -- will add a break

Reply:
added a commit on top